server/src/main/java/io/crate/execution/engine/pipeline/ProjectionToProjectorVisitor.java

```diff
@@ -554,6 +557,21 @@ public Projector visitUpdateProjection(final UpdateProjection projection, Contex
     private <A> ShardDMLExecutor<?, ?, A, ?> buildUpdateShardDMLExecutor(
         Context context, UpdateProjection projection,
         Collector<ShardResponse, A, Iterable<Row>> collector) {
+        // Get Stats to improve ram estimation for the update items
+        final Stats stats;
+        final Collection<Reference> ramAccountedColumns;
+        assert shardId != null : "ShardId must be provided for updates";
+        String indexName = shardId.getIndexName();
+        RelationName relationName = RelationName.fromIndexName(indexName);
+        stats = nodeCtx.tableStats().getStats(relationName);
+        // if stats are not available we fall back to estimate the size based on
+        // column types. Therefore we need to get the column information.
+        if (stats.isEmpty()) {
+            TableInfo tableInfo = nodeCtx.schemas().getTableInfo(relationName);
+            ramAccountedColumns = tableInfo.allColumns();
+        } else {
+            ramAccountedColumns = List.of();
```

Comment:
This is done now at 2 places, spawning its logic across multiple components. why not simple calculate the additional bytes here already and pass it in?

Reply:
At this point, there are no items and no requests. This just initializes the stats and the columns to be created once the items are added. The alternative would be to move this logic into `ShardUpsertRequest.Item.forUpdate`, but then it would be called on each added item.

Reply:
How about the following design?

Reply:
I still don't understand why it is better when split across 2 places, why not:
```suggestion
        // if stats are not available we fall back to estimate the size based on
        // column types. Therefore we need to get the column information.
        long sizeEstimatePerRow;
        if (stats.isEmpty()) {
            TableInfo tableInfo = nodeCtx.schemas().getTableInfo(relationName);
            sizeEstimatePerRow = stats.estimateSizeForColumns(tableInfo.allColumns());
        } else {
            sizeEstimatePerRow = stats.averageSizePerRowInBytes();
```

Reply:
It is hard to test when it's hardcoded here. It was there before, but @mfussenegger suggested moving it into `ShardUpsertRequest` to make it easy to test.

Reply:
I see, but this is not completely moving it into the request but splitting the logic instead imho.
Can't we pass in the `DocTable` and `Stats` into the request, or at least the static function, and by this, remove the 2nd if/else?

Reply:
Sounds good, thanks for the idea.